File: pipeline_scripts/agglomerators/universal_community_agglomerate.R
# Agglomerates by comunnity
options(warn=-1)

# Imports the libraries
suppressMessages(library("dplyr"))
suppressMessages(library('igraph'))
suppressMessages(library('proxy'))
suppressMessages(library('ramify'))


source("pipeline_scripts/functions/agglomeration_functions.R")

# Directories
source("global_config/config.R")
data_dir = get_property('data_dir')
analysis_dir = get_property('analysis_dir')


# Ident for console
ident = '         '


args = commandArgs(trailingOnly=TRUE)

# For Debug
debug = FALSE
if(debug)
{   
  cat("\n\n\n\n\n\n\n\n\n\n\n¡¡¡¡¡DEBUG IS ON!!!!\n\n\n\n\n\n\n\n\n")
  setwd("~/Dropbox/Projects/covid_fb_pipeline/covid_geoinsights_pipeline")
  args = c('Peru','peru')
}


location_name = args[1] # Location Name
location_folder = args[2] # Location Folder
agglomeration_method = args[3] # Agglomeration method to build on to



if(is.na(agglomeration_method))
{
  # First tries geometry
  if(dir.exists(file.path(data_dir, 'data_stages',location_folder,'agglomerated','geometry')))
    agglomeration_method = 'geometry'
  else if(dir.exists(file.path(data_dir, 'data_stages',location_folder,'agglomerated','radial')))
    agglomeration_method = 'radial'
  else
    stop('No agglomeration method found to build communiies')
}



cat(paste(ident, 'Community Agglomeration for: ', location_name,  '\n', sep = ""))

# Agglomerated Folder
agglomerated_folder = file.path(data_dir, 'data_stages',location_folder,'agglomerated', agglomeration_method)

# Export Folder
# Two steps
export_folder = file.path(data_dir, 'data_stages',location_folder, 'agglomerated')
dir.create(export_folder)

export_folder = file.path(export_folder, 'community')
dir.create(export_folder)



# Loads movement
cat(paste(ident, '   Loads Movement','\n', sep = ""))
movement = read.csv(file.path(agglomerated_folder, 'movement.csv'),  stringsAsFactors = FALSE)
polygons = read.csv(file.path(agglomerated_folder, 'polygons.csv'),  stringsAsFactors = FALSE)
cases = read.csv(file.path(agglomerated_folder, 'cases.csv'), stringsAsFactors = FALSE)

# Extracts the cases columns
cases_col = colnames(cases)[as.vector(sapply(colnames(cases), function(col){grepl('num_', col, fixed = TRUE)}))]
# Extracts the attr columns
attr_col = colnames(polygons)[as.vector(sapply(colnames(polygons), function(col){grepl('attr_', col, fixed = TRUE)}))]


# Groups the dataframe
g_mov = movement %>% 
        group_by(start_poly_id, end_poly_id) %>% 
        summarise(movement = mean(movement), .groups = "keep") %>%
        ungroup() %>%
        select(start_poly_id, end_poly_id, movement)

# Creates the graph
cat(paste(ident, '   Creates the Graph','\n', sep = ""))
g <- graph_from_data_frame(d = g_mov %>% rename(weight = movement), directed = FALSE, vertices = polygons$poly_id)
g = simplify(g)


cat(paste(ident, '   Finds Communities and Pagerank','\n', sep = ""))
# Clusters
wc <- cluster_walktrap(g)

# Assigns the community for each center
polygons$community_id = membership(wc)


cat(paste(ident, '   Agglomerates Cases','\n', sep = ""))


# creates the agglomerated polygons
#------------------------
cat(paste(ident, '   Agglomerates Polygons','\n', sep = ""))

# Final_comunity
# I fpopulation is found it will be used for the id
# if not, the number of cases will be used
if('attr_population' %in% colnames(polygons))
{
  final_community = polygons %>%
                    select(poly_id, community_id, attr_population) %>%
                    group_by(community_id) %>%
                    summarise(final_id = exctract_id_by_population(poly_id, attr_population), .groups = "keep") %>%
                    ungroup()  
}else
{
    final_community = polygons %>%
                    select(poly_id, community_id, num_cases) %>%
                    group_by(community_id) %>%
                    summarise(final_id = exctract_id_by_cases(poly_id, num_cases), .groups = "keep") %>%
                    ungroup()  
}


polygons = polygons %>%
           inner_join(final_community, by = c('community_id' = 'community_id')) %>%
           mutate(community_id = final_id)



# Creates the community name:
# If population exists it uses it if not, number of cases
if('attr_population' %in% colnames(polygons))
{
    agg_poly_1 = polygons %>% 
    group_by(community_id) %>%
    summarise(poly_name = exctract_name_by_population(poly_name, attr_population), agglomerated_polygons = extract_list_of_agg_polygons(poly_id), geometry = extract_geometry(geometry), poly_lon = extract_center_by_population(poly_lon, poly_lat, attr_population)[1], poly_lat = extract_center_by_population(poly_lon, poly_lat, attr_population)[2], .groups = "keep") %>%
    ungroup()  
  
}else
{
  agg_poly_1 = polygons %>% 
  group_by(community_id) %>%
  summarise(poly_name = exctract_name_by_cases(poly_name, num_cases), agglomerated_polygons = extract_list_of_agg_polygons(poly_id), geometry = extract_geometry(geometry), poly_lon = extract_center_by_cases(poly_lon, poly_lat, num_cases)[1], poly_lat = extract_center_by_cases(poly_lon, poly_lat, num_cases)[2], .groups = "keep") %>%
  ungroup()  
}


# Creates the location to community map
polygon_community_map = agg_poly_1 %>% 
                        select(community_id, poly_name) %>%
                        rename(community_name = poly_name) %>%
                        inner_join(polygons, by = c('community_id' = 'final_id')) %>%
                        select(poly_id, poly_name, community_id, community_name )
          


agg_poly = polygons %>% 
  group_by(community_id) %>%
  summarise_at(c(attr_col, cases_col), sum, na.rm = TRUE) %>%
  ungroup() %>%
  inner_join(agg_poly_1, by = c('community_id' = 'community_id')) %>%
  select(c('community_id', 'poly_name', 'agglomerated_polygons', 'geometry', 'poly_lon', 'poly_lat', all_of(c(attr_col,cases_col)))) %>%
  rename(poly_id = community_id) %>%
  arrange(desc(num_cases))



# Creates the agglomerated cases
# ------------------------------
agg_cases = cases %>% 
            inner_join(polygons %>% select(poly_id,  community_id), by = c('poly_id' = 'poly_id')) %>%
            group_by(date_time, community_id)  %>%
            summarise_at(cases_col, sum, na.rm = TRUE) %>%
            ungroup() %>%
            inner_join(agg_poly %>% select(poly_id, poly_name) , by = c('community_id' = 'poly_id')) %>%
            select(c('date_time', 'community_id', 'poly_name', all_of(cases_col))) %>%
            rename(poly_id = community_id,location = poly_name) %>%
            arrange(date_time, desc(num_cases))
            



# creates the agglomerated Movement
#------------------------
cat(paste(ident, '   Agglomerates Movement','\n', sep = ""))
agg_mov = movement %>% 
          inner_join(polygons %>% select(poly_id, community_id), by = c('start_poly_id' = 'poly_id')) %>%
          mutate(start_poly_id = community_id) %>%
          select(date_time, start_poly_id, end_poly_id, movement) %>%
          inner_join(polygons %>% select(poly_id, community_id), by = c('end_poly_id' = 'poly_id')) %>%
          mutate(end_poly_id = community_id) %>%
          select(date_time, start_poly_id, end_poly_id, movement) %>%
          group_by(date_time, start_poly_id, end_poly_id) %>%
          summarise(movement = sum(movement), .groups = "keep") %>%
          ungroup() %>%
          arrange(date_time)



# creates the agglomerated Populations
#------------------------
cat(paste(ident, '   Agglomerates Population','\n', sep = ""))
# Loads the cases data to assign the comunity ids and name them
pop = read.csv(file.path(agglomerated_folder, 'population.csv'), stringsAsFactors = FALSE)


if(nrow(pop) > 0)
{
  
  agg_pop = pop %>%
    inner_join(polygons %>% select(poly_id, community_id), by = c('poly_id' = 'poly_id')) %>%
    mutate(poly_id = community_id) %>%
    select(date_time, poly_id, population) %>%
    group_by(date_time, poly_id) %>%
    summarise(population = sum(population), .groups = "keep") %>%
    ungroup() %>%
    arrange(date_time)
}else
{
  agg_pop = pop
}




# Checks if movent range exits
agg_mov_range = NULL
movement_range_file = file.path(agglomerated_folder,'movement_range.csv')
if(file.exists(movement_range_file))
{
  df_movement_range = read.csv(movement_range_file)
  
  filtered_polys = polygons[,c('poly_id','community_id')]
  filtered_polys$factor = 1
  
  'attr_population' %in% colnames(polygons)
  
  if('attr_density' %in% colnames(polygons))
  {
    filtered_polys$factor = polygons$attr_density
    
  }else if('attr_population' %in% colnames(polygons) && 'attr_area' %in% colnames(polygons))
  {
    filtered_polys$factor = polygons$attr_population / polygons$attr_area
    
  }else if('attr_population' %in% colnames(polygons))
  {
    filtered_polys$factor = polygons$attr_population
    
  }else if('attr_area' %in% colnames(polygons))
  {
    filtered_polys$factor = polygons$attr_area    
  }
  
  agg_mov_range = df_movement_range %>% 
                  inner_join(filtered_polys, by = c('poly_id' = 'poly_id')) %>%
                  group_by(date_time, community_id)  %>%
                  summarise(movement_change = factor_average(movement_change, factor), .groups = "keep") %>%
                  ungroup() %>%
                  select(c('date_time', 'community_id', 'movement_change')) %>%
                  rename(poly_id = community_id) %>%
                  arrange(date_time)
                  
}



cat(paste(ident, '   Saves Files','\n', sep = ""))
# Saves all the datasets
write.csv( agg_cases, file.path(export_folder,'cases.csv'), row.names = FALSE)
write.csv( agg_mov, file.path(export_folder,'movement.csv'), row.names = FALSE)
write.csv(agg_poly, file.path(export_folder,'polygons.csv'), row.names = FALSE)
write.csv(agg_pop, file.path(export_folder,'population.csv'), row.names = FALSE)
write.csv(polygon_community_map, file.path(export_folder,'polygon_community_map.csv'), row.names = FALSE)

if(!is.null(agg_mov_range))
{
  write.csv(agg_mov_range, file.path(export_folder,'movement_range.csv'), row.names = FALSE)
}

cat(paste(ident, 'Done!','\n', sep = ""))

File: pipeline_scripts/agglomerators/universal_community_agglomerate.py
# Universal Community Agglomeration script
# This script agglomerates a certain other agglomeration using the walking trap algorithm. 
# This implementation currently excecutes an outside R script for the polygon to community map.
# The attributes are agglomerated acording to the impemented attribute scheme (created in the unify step)

# Imports the necessary libraries
import pandas as pd
import numpy as np
import os
import sys

# Excecution Functions
import excecution_functions as ef

# Attribute agglomerator 
from attr_agglomeration_functions import *

# Directories
from global_config import config
data_dir = config.get_property('data_dir') # Data Directory

# Ident
ident = '      '

# Reads the parameter inputs
location_name  = sys.argv[1] # location name
location_folder_name  = sys.argv[2] # location folder name
source_agglomeration  = sys.argv[3] # Type of agglomeration to build upon (source)

# Debug

#location_name  = 'Colombia'
#location_folder_name  = 'colombia'
#source_agglomeration  = 'geometry'


# Agglomeration name
agglomeration_method = 'community' # Constant

print(ident + f'Computing Community agglomeration for: {location_name} over: {source_agglomeration}')
      
# Sets the location
location_folder = os.path.join(data_dir, 'data_stages', location_folder_name)

# Source and destination agglomeration
source_agglomeration_folder = os.path.join(location_folder, 'agglomerated/', source_agglomeration)
destination_agglomeration_folder = os.path.join(location_folder, 'agglomerated/', agglomeration_method)

# Checks if the agglomeration to build upon exists
if not os.path.exists(source_agglomeration_folder):
    raise ValueError(f'Agglomeration method: {source_agglomeration} does not exist for: {location_name}. Please compute it!')

# Creates the destination agglomeration folder
if not os.path.exists(destination_agglomeration_folder):
    os.makedirs(destination_agglomeration_folder)


# 1 -- Computes the Community map (Calls R script)
print(ident + '   Computing the Community Map (R Code)')
script_location = 'pipeline_scripts/agglomerators/'
name = 'compute_community_agglomeration_map'
code_type = 'R'
parameters = f'{location_name} {location_folder_name} {source_agglomeration}'

# Excecutes Script
resp = ef.excecute_script(script_location, name, code_type, parameters)

if resp != 0: # Error
    raise ValueError('Error when computing the Community Map. Will not agglomerate.')
    

# 2 -- Loads all the necessary files
print(ident + '   Loads Files')
# Cases
df_cases = pd.read_csv(os.path.join(source_agglomeration_folder, 'cases.csv'), parse_dates = ['date_time'], dtype = {'poly_id':str})

# Movement
df_movement = pd.read_csv(os.path.join(source_agglomeration_folder, 'movement.csv'), parse_dates = ['date_time'], dtype = {'start_poly_id':str, 'end_poly_id':str})

# Polygons
df_polygons = pd.read_csv(os.path.join(source_agglomeration_folder, 'polygons.csv'), dtype = {'poly_id':str})

# Population (Mock)
df_population = pd.read_csv(os.path.join(source_agglomeration_folder, 'population.csv'), parse_dates = ['date_time'], dtype = {'poly_id':str})

# Movement Range (if Exists)
df_movement_range = None
if os.path.exists(os.path.join(source_agglomeration_folder, 'movement_range.csv')):
    df_movement_range = pd.read_csv(os.path.join(source_agglomeration_folder, 'movement_range.csv'), parse_dates = ['date_time'], dtype = {'poly_id':str})

# Agglomeration Scheme
aggl_scheme = pd.read_csv(os.path.join(data_dir, 'data_stages', location_folder_name, 'unified/aggl_scheme.csv'))

# Community map
community_map = pd.read_csv(os.path.join(destination_agglomeration_folder, 'polygon_community_map.csv'), dtype = {'poly_id':str, 'community_id':str})

community_map = community_map[["poly_id","community_id","community_name"]].copy()



# 3 -- Agglomerates
print(ident + '   Agglomerates')

# Cases
# -----------
print(ident + '      Cases')
# Merges
df_cases_final = df_cases.merge(community_map, on = 'poly_id', how = 'left')

# Drops and renames
df_cases_final.drop(['poly_id','location'], axis = 1, inplace = True)
df_cases_final.rename(columns = {'community_id':'poly_id', 'community_name':'location'}, inplace = True)

# Groups by
groupby_cols = ['date_time','location','poly_id']
agglomerate_cols = df_cases_final.columns.drop(groupby_cols).values
df_cases_final = agglomerate(df_cases_final, aggl_scheme, groupby_cols, agglomerate_cols)


# Movement
# ---------------
print(ident + '      Movement')

# Merges
# Start id
df_movement_final = df_movement.merge(community_map[['poly_id','community_id']].rename(columns = {'poly_id':'start_poly_id'}), on = 'start_poly_id', how = 'left')
df_movement_final.drop(['start_poly_id'], axis = 1, inplace = True)
df_movement_final.rename(columns = {'community_id':'start_poly_id'}, inplace = True)

# End id
df_movement_final = df_movement_final.merge(community_map[['poly_id','community_id']].rename(columns = {'poly_id':'end_poly_id'}), on = 'end_poly_id', how = 'left')
df_movement_final.drop(['end_poly_id'], axis = 1, inplace = True)
df_movement_final.rename(columns = {'community_id':'end_poly_id'}, inplace = True)

# Groups by
groupby_cols = ['date_time','start_poly_id','end_poly_id']
agglomerate_cols = df_movement_final.columns.drop(groupby_cols).values
df_movement_final = agglomerate(df_movement_final, aggl_scheme, groupby_cols, agglomerate_cols)


# Polygons
# ------------
print(ident + '      Polygons')

# Merges
df_polygons_final = df_polygons.merge(community_map, on = 'poly_id', how = 'left')

# Drops and renames
df_polygons_final.drop(['poly_id','community_name'], axis = 1, inplace = True)
df_polygons_final.rename(columns = {'community_id':'poly_id'}, inplace = True)

# Groups by
groupby_cols = ['poly_id']
agglomerate_cols = df_polygons_final.columns.drop(groupby_cols).values
df_polygons_final = agglomerate(df_polygons_final, aggl_scheme, groupby_cols, agglomerate_cols)


# Population
# ---------------
print(ident + '      Population')

# Merges
df_population_final = df_population.merge(community_map[['poly_id','community_id']], on = 'poly_id', how = 'left')

# Drops and renames
df_population_final.drop(['poly_id'], axis = 1, inplace = True)
df_population_final.rename(columns = {'community_id':'poly_id'}, inplace = True)


# Groups by (Mock)
#groupby_cols = ['date_time','poly_id']
#agglomerate_cols = df_population_final.columns.drop(groupby_cols).values
#df_population_final = agglomerate(df_population_final, aggl_scheme, groupby_cols, agglomerate_cols)


# Movement Range
# ---------------
# Does the procedure manually, since the agglomeration does not have support for columnes outside the 
# dataframe.

df_movement_range_final = None
if df_movement_range is not None:
    
    print(ident + '      Movement Range')
    # Merges with original polygons
    df_movement_range_final = df_movement_range.merge(df_polygons, on = 'poly_id', how = 'left')

    # Merges with community map
    df_movement_range_final = df_movement_range_final.merge(community_map[['poly_id','community_id']], on = 'poly_id', how = 'left')

    # Drops and renames
    df_movement_range_final.drop(['poly_id'], axis = 1, inplace = True)
    df_movement_range_final.rename(columns = {'community_id':'poly_id'}, inplace = True)

    groupby_cols = ['date_time','poly_id']
    agglomerate_cols = ['movement_change']

    # Default option is average
    temp_scheme_dic = {'attr_name':'movement_change', 'aggl_function':'attr_average', 'secondary_attr': None, 'aggl_parameters': None}

    # Checks options
    if "attr_population" in df_movement_range_final: # By population
        temp_scheme_dic['aggl_function'] = 'attr_weighted_average'
        temp_scheme_dic['secondary_attr'] = 'attr_population'

    elif "attr_area" in df_movement_range_final: # By Area
        temp_scheme_dic['aggl_function'] = 'attr_weighted_average'
        temp_scheme_dic['secondary_attr'] = 'attr_area'

    # Declares Agglomeration Scheme
    temp_scheme = pd.DataFrame(temp_scheme_dic, index = [0])
    # Groups by
    df_movement_range_final = agglomerate(df_movement_range_final, temp_scheme, groupby_cols, agglomerate_cols)

    
# 4 -- Saves the dataframes
print(ident + '   Saves')
# Cases
df_cases_final.to_csv(os.path.join(destination_agglomeration_folder, 'cases.csv'), index = False)
# Movement
df_movement_final.to_csv(os.path.join(destination_agglomeration_folder, 'movement.csv'), index = False)
# Population
df_population_final.to_csv(os.path.join(destination_agglomeration_folder, 'population.csv'), index = False)
# Polygons
df_polygons_final.to_csv(os.path.join(destination_agglomeration_folder, 'polygons.csv'), index = False)

# Movement Change
if df_movement_range_final is not None:
    df_movement_range_final.to_csv(os.path.join(destination_agglomeration_folder, 'movement_range.csv'), index = False)
    
    
print(ident + 'Done')    